lib/downloader.js

```diff
@@ -76,13 +76,38 @@ function download (torrentId) {
       torrent.addWebSeed(webSeedUrl)
     }
 
+    torrentMeta = parseTorrent(torrent.torrentFile)
+
+    if (!priority) {
+      priority = new IdbKvStore('planktos-priority')
+      priority.on('add', function (change) {
+        onPriorityAdd(change.value)
+      })
+    }
+
+    priority.values().then(values => values.forEach(v => onPriorityAdd(v)))
+
     torrent.on('done', function () {
       debug('TORRENT DOWNLOADED', torrent.files.map(f => f.name))
-      let channel = new BroadcastChannel('planktos-downloaded')
-      torrent.files.forEach(function (f) {
-        downloaded.set(f.name, true)
-        .then(() => channel.postMessage({ name: f.name }))
-      })
+      priority.clear()
     })
   })
 }
+
+function onPriorityAdd (hash) {
+  let fileInfo = torrentMeta.files.find(f => f.name === hash)
+  let start = Math.floor(fileInfo.offset / torrentMeta.pieceLength)
+  let end = Math.floor((fileInfo.offset + fileInfo.length - 1) / torrentMeta.pieceLength)
+  let torrent = webtorrent.get(torrentMeta.infoHash)
+  let file = torrent.files.find(f => f.name === hash)
+
+  if (file.downloaded !== file.length) {
+    torrent.select(start, end, 1)
+  } else {
+    var transaction = priority.transaction()
+    transaction.json().then(json => {
+      var rm = Object.keys(json).filter(k => json[k] === hash)
+      rm.forEach(k => transaction.remove(k))
```

Comment:
Related to your comment on downloader.js#66 . Only the downloader knows if the file is downloaded so whenever a file is requested it's always added to the priority store.

Maybe re-adding the 'downloaded' store would be better. But then we get inconsistencies between what the 'downloaded' store says is downloaded and what is actually downloaded (at least for a short time period). Alternatively maybe there is a way to look at the chunk-store backed idb-kv-store and see if the keys for the file's pieces are all present. This would prevent the inconsistencies, but some code would have to be added to idb-kv-store for this to work.